src/gevent/_hub_primitives.py

```diff
@@ -235,7 +246,9 @@ def wait_on_objects(objects=None, timeout=None, count=None):
     if objects is None:
         hub = get_hub()
         return hub.join(timeout=timeout) # pylint:disable=
-    return list(iwait_on_objects(objects, timeout, count))
+
+    with iwait_on_objects(objects, timeout, count) as iterator:
```

Comment:
Why do we need this change? `list` will either consume the iterator, or an exception will be raised and the iterator cleans itself up.